traffic_portal/app/src/common/modules/form/topology/FormTopologyController.js

```diff
@@ -0,0 +1,258 @@
+
+	let cacheGroupNamesInTopology = [];
+
+	let hydrateTopology = function() {
+		// add some needed fields to each cache group (aka node) of a topology
+		topology.nodes.forEach(function(node) {
+			let cacheGroup = cacheGroups.find( function(cg) { return cg.name === node.cachegroup} );
+			Object.assign(node, { id: cacheGroup.id, type: cacheGroup.typeName });
+		});
+	};
+
+	let removeSecParentReferences = function(topologyTree, secParentName) {
+		// when a cache group is removed, any references to the cache group as a secParent need to be removed
+		topologyTree.forEach(function(node) {
+			if (node.secParent && node.secParent === secParentName) {
+				node.secParent = '';
+			}
+			if (node.children && node.children.length > 0) {
+				removeSecParentReferences(node.children, secParentName);
+			}
+		});
+	};
+
+	// build a list of cache group names currently in the topology
+	let buildCacheGroupNamesInTopology = function(topologyTree, fromScratch) {
+		if (fromScratch) cacheGroupNamesInTopology = [];
+		topologyTree.forEach(function(node) {
+			if (node.cachegroup) {
+				cacheGroupNamesInTopology.push(node.cachegroup);
+			}
+			if (node.children && node.children.length > 0) {
+				buildCacheGroupNamesInTopology(node.children, false);
+			}
+		});
+	};
+
+	$scope.topology = topology;
+
+	$scope.topologyTree = [];
+
+	$scope.topologyTreeOptions = {
+		beforeDrop: function(evt) {
+			let node = evt.source.nodeScope.$modelValue,
+				parent = evt.dest.nodesScope.$parent.$modelValue;
+
+			if (!parent || !node) {
+				return false; // no dropping outside the toplogy tree and you need a node to drop
+			}
+
+			// ORG_LOC cannot have a parent
+			if (node.type === 'ORG_LOC' && parent.cachegroup) {
+				$anchorScroll(); // scrolls window to top
+				messageModel.setMessages([ { level: 'error', text: 'Cache groups of ORG_LOC type must not have a parent.' } ], false);
+				return false;
+			}
+
+			// EDGE_LOC cannot have children
+			if (parent.type === 'EDGE_LOC') {
+				$anchorScroll(); // scrolls window to top
+				messageModel.setMessages([ { level: 'error', text: 'Cache groups of EDGE_LOC type must not have children.' } ], false);
+				return false;
+			}
+
+			// update the parent and secParent fields of the node on successful drop
+			if (parent.cachegroup) {
+				node.parent = parent.cachegroup; // change the node parent based on where the node is dropped
+				if (node.parent === node.secParent) {
+					// node parent and secParent cannot be the same
+					node.secParent = "";
+				}
+			} else {
+				// the node was dropped at the root of the topology. no parents.
+				node.parent = "";
+				node.secParent = "";
+			}
+			return true;
+		}
+	};
+
+	$scope.navigateToPath = locationUtils.navigateToPath;
+
+	$scope.hasError = formUtils.hasError;
+
+	$scope.hasPropertyError = formUtils.hasPropertyError;
+
+	$scope.nodeLabel = function(node) {
+		if (!node.cachegroup) return 'TOPOLOGY';
+		return node.cachegroup;
```

Comment:
I think this could be `return node.cachegroup || 'TOPOLOGY';` without needing to use a weird one-line `if`

Reply:
you don't like the one line IFs? I thought they were cool. :)

Reply:
I think they're hard to read since ifs usually enclose a block